src/modules/rangeproof/rangeproof_impl.h

```diff
@@ -59,188 +257,196 @@ SECP256K1_INLINE static void secp256k1_rangeproof_serialize_point(unsigned char*
     secp256k1_fe_get_b32(data + 1, &pointx);
 }
 
-SECP256K1_INLINE static int secp256k1_rangeproof_genrand(secp256k1_scalar *sec, secp256k1_scalar *s, unsigned char *message,
- size_t *rsizes, size_t rings, const unsigned char *nonce, const secp256k1_ge *commit, const unsigned char *proof, size_t len, const secp256k1_ge* genp) {
-    unsigned char tmp[32];
+SECP256K1_INLINE static void secp256k1_rangeproof_init_rng(
+    secp256k1_rfc6979_hmac_sha256* rng,
+    const unsigned char* nonce,
+    const secp256k1_ge* commit,
+    const unsigned char *proof,
+    const size_t len,
+    const secp256k1_ge* genp
+) {
     unsigned char rngseed[32 + 33 + 33 + 10];
-    secp256k1_rfc6979_hmac_sha256 rng;
-    secp256k1_scalar acc;
-    int overflow;
-    int ret;
-    size_t i;
-    size_t j;
-    int b;
-    size_t npub;
     VERIFY_CHECK(len <= 10);
+
     memcpy(rngseed, nonce, 32);
     secp256k1_rangeproof_serialize_point(rngseed + 32, commit);
     secp256k1_rangeproof_serialize_point(rngseed + 32 + 33, genp);
     memcpy(rngseed + 33 + 33 + 32, proof, len);
-    secp256k1_rfc6979_hmac_sha256_initialize(&rng, rngseed, 32 + 33 + 33 + len);
-    secp256k1_scalar_clear(&acc);
+    secp256k1_rfc6979_hmac_sha256_initialize(rng, rngseed, 32 + 33 + 33 + len);
+}
+
+SECP256K1_INLINE static int secp256k1_rangeproof_genrand_sign(
+    secp256k1_scalar *sec,
+    secp256k1_scalar *s,
+    const unsigned char* blind,
+    const unsigned char* message,
+    const size_t msg_len,
+    const secp256k1_rangeproof_header* header,
+    const uint64_t proven_value,
+    secp256k1_rfc6979_hmac_sha256* rng
+) {
+    unsigned char tmp[32];
+    secp256k1_scalar acc;
+    int overflow;
+    int ret;
+    size_t i;
+    size_t npub;
+    secp256k1_scalar_set_b32(&acc, blind, &overflow);
+    if (overflow) {
+        return 0;
+    }
+    secp256k1_scalar_negate(&acc, &acc);
     npub = 0;
     ret = 1;
-    for (i = 0; i < rings; i++) {
-        if (i < rings - 1) {
-            secp256k1_rfc6979_hmac_sha256_generate(&rng, tmp, 32);
+    for (i = 0; i < header->n_rings; i++) {
+        size_t j;
+        if (i < header->n_rings - 1) {
+            secp256k1_rfc6979_hmac_sha256_generate(rng, tmp, 32);
             do {
-                secp256k1_rfc6979_hmac_sha256_generate(&rng, tmp, 32);
+                secp256k1_rfc6979_hmac_sha256_generate(rng, tmp, 32);
                 secp256k1_scalar_set_b32(&sec[i], tmp, &overflow);
             } while (overflow || secp256k1_scalar_is_zero(&sec[i]));
             secp256k1_scalar_add(&acc, &acc, &sec[i]);
         } else {
             secp256k1_scalar_negate(&acc, &acc);
             sec[i] = acc;
```

Comment:
I am actually thinking about removing all of the checks for conditions like this from the signing code, since they are cryptographically impossible to hit.